[Doc] QueryInsightsDashboard (#9157)
opensearch-project/documentation-website: _observing-your-data/query-insights/query-insights-dashboard.md

You can set the following specific grouping attributes using these steps.

1. Select a grouping option under **Group By**, such as **Similarity**.
2. Select **Save**.
3. Check the **Statues for group by** panel to see if **Group by** criteria is enabled.

### Data export and retention

To configuring data export and retention, use the **Query insights export and data retention settings** panel. There, you can set the following settings:

1. Under **Exporter**, choose a destination where to export the insight data, such as **Local index**.
2. Set the **Delete After (days)** field with a data retention period.
3. Select **Save**.
4. In the **Statuses for data retention** panel, make sure that the **Exporter** setting is enabled.

### Configuration best practices

When configuring the query insights feature, remember the following best practices:

- Begin with a smaller value for N (count) and increase it based on your system's load.
- When setting the retention periods for insight data, consider a shorter retention period that saves storage but reduces long-term insights.
- Enable metrics based on your monitoring needs. Monitoring fewer metrics prevents system overload.

> - Configure a larger time window for more comprehensive insights.

Reviewer: Are we referring to a larger data retention window here @ansjcy and @KishoreKicha14?

Author: For analyzing queries. I have rephrased the sentence.

Reviewer 2: No, and I don't think 
> Set a larger time window to analyze queries over a longer duration for more comprehensive insights.

is accurate neither, a larger time window means larger granularity, but can significantly save the compute resource. For example, gathering the top n for each 1 minute / top n for each day. But it's definitely not `more comprehensive`

@KishoreKicha14 Could you update this? Please refer to the existing API doc https://opensearch.org/docs/latest/observing-your-data/query-insights/top-n-queries/#configuring-the-window-size when you write the dashboards doc.

---
title: Query insights dashboards
layout: default
parent: Query insights
nav_order: 60
---

# Query insights dashboards

> You can interact with the query insights feature using the Query Insights Dashboards plugin. This plugin gives you real-time insights into query performance, providing analytics, and monitoring to improve how queries are run in your cluster.

Reviewer 2: Should we also say historical insights? i.e real time and historical insights?

> You can interact with the query insights feature using the Query Insights Dashboards plugin. This plugin gives you real-time  nd historical insights into query performance, providing analytics, and monitoring to improve how queries are run in your cluster.

Reviewer: ```suggestion
You can interact with the query insights feature using the Query Insights Dashboards plugin. This plugin gives you real-time and historical insights into query performance, providing analytics, and monitoring to improve how queries are run in your cluster.
```

> You can interact with the query insights feature using the Query Insights Dashboards plugin. This plugin gives you real-time  and historical insights into query performance, providing analytics, and monitoring to improve how queries are run in your cluster.

Reviewer: ```suggestion
You can interact with the query insights feature using the Query Insights Dashboards plugin. This plugin gives you real-time  nd historical insights into query performance, providing analytics, and monitoring to improve how queries are run in your cluster.
```

## Prerequisites

The Query Insights Dashboards plugin requires [OpenSearch 2.19 or later]({{site.url}}{{site.baseurl}}/install-and-configure/). 

## Installing the plugin

To install the Query Insights Dashboards plugin, see [Managing OpenSearch Dashboards plugins]({{site.url}}{{site.baseurl}}/install-and-configure/install-dashboards/plugins/).

## Navigation

After logging in to OpenSearch Dashboards, you can find the Query Insights page by navigating to **OpenSearch Plugins** > **Query Insights**. 

> 

Reviewer: Confirmed

![Top N Queries Interface](../../images/Query-Insights/QueryInsights.png)

Each label corresponds with the following components:

- [A. Navigation tabs](#a-navigation-tabs)
- [B. Search queries bar](#b-search-queries-bar)
- [C. Filters](#c-filters)
- [D. Date range selector](#d-date-range-selector)
- [E. Refresh button](#e-refresh-button)
- [F. Metrics table](#f-metrics-table)

### A. Navigation tabs

The navigation tabs allow you to switch between the Configuration and Top N Queries pages.

### B. Search queries bar

The search queries bar filters queries based on specific attributes such as **query type** or **indexes**. You can use additional filters as shown in the [filters](#c-filters) section.

### C. Filters

The filters dropdown menus allow you to select the following query filters.

| Filter                  | Description                                                         | Example            |
|-------------------------|---------------------------------------------------------------------|--------------------|
| **Type**                | Filter by query type.                                               | `query`, `group`   |

> | **Indexes**             | Filter queries based on specific OpenSearch indices.                | `index1`, `index2` |

Reviewer: ```suggestion
| **Indexes**             | Filter queries based on specific OpenSearch indexes.                | `index1`, `index2` |
```

| **ID**                  | The unique identifier for the query.                                         |
| **Type**                | The type of query, such as `query`or `group`.                               |
| **Query Count**         | The number of times that the query has been executed.                            |
| **Timestamp**           | When the query was run.                                                     |
| **Latency**             | The time taken for the query to execute.                                    |
| **CPU Time**            | The CPU resources consumed by the query.                                    |
| **Memory Usage**        | The memory usage of the query.                                              |
| **Indexes**             | The index or indexes on which the query was executed.                       |
| **Search Type**         | The type of search used, for example, `query then fetch`.                   |
| **Coordinator Node ID** | The node that coordinated the query.                                        |
| **Total Shards**        | The total number of shards involved in running the query.                    |

## Query details

> The **Query details** page provides insights into query behavior, performance, and structure. You can access the query details page by selecting the query ID, as shown in the following image.

Reviewer: ```suggestion
The **Query details** page provides insights into query behavior, performance, and structure. You can access the query details page by selecting the query ID, as shown in the following image:
```

> ![Query Insights List](../../images/Query-Insights/Querieslist.png)

Reviewer: @natebower: Adjusted the language.

The **Query Details** page in OpenSearch Dashboards provides insights into query behavior, performance, and structure. This page can have two variations based on the context:

![Query Insights List](../../images/Query-Insights/Querieslist.png)

### Viewing individual query details

You can access detailed information about a single query by select the Query ID, such as `51c68a1a-7507-4b3e-aea1-32ddd74dbac4`. The query details page will appear, as shown in the following image:

![Individual Query Details](../../images/Query-Insights/IndividualQueryDetails.png)

In the query details view, you can view the following information such as **Timestamp**, **CPU Time**, **Memory Usage**, **indexes**, **Search Type**, **Coordinator Node ID**, and **Total Shards**.

> ### Viewing query group details

Reviewer 2: @KishoreKicha14  Could you attach the screenshot here? (GroupQueryDetails.png)

> To view query group details, select a query ID marked as a "group" in the **Top N queries** list. The query group details view provides the information shown in the following image and listed below.

Reviewer: ```suggestion
To view query group details, select a query ID marked as a "group" in the **Top N queries** list. The query group details view provides the following information:
```

![Query Group Details](../../images/Query-Insights/GroupQueryDetails.png)

- The **Aggregate summary for Queries** section, gives a view of key query metrics for the entire group, including **Average latency**, **Average CPU time**, **Average memory usage**, and, **Group by** criteria.
- The **Sample query details** section provides information about a single representative query including its **Timestamp**, **Indexes**, **Search Type**, **Coordinator Node ID**, and the **Total Shards**.
- The **Query** section displays the JSON structure of the query. 

> - The **Latency Breakdown** section presents a graphical representation of the run phases for the query.

Reviewer: Correct.

> - The **Latency Breakdown** section presents a graphical representation of the run phases for the query.

Reviewer: ```suggestion
- The **Latency** section presents a graphical representation of the run phases for the query.
```

- The **Latency Breakdown** section presents a graphical representation of the run phases for the query.

## Configuration

The **Query insights - Configuration** page is designed to gives you control over how query insights collects, monitors, groups, and retains data. The following image shows the configuration page:

![Configuration](../../images/Query-Insights/Configuration.png)

On the page, you can configure the following settings.

### Top N queries monitoring

The **Top n queries monitoring configuration settings** allows you to track query performance metrics, such as **Latency, CPU Usage, and Memory**, to analyze and optimize query performance. The configuration interface provides a structured, menu-driven setup where you can define the specific metrics to monitor, set thresholds for analysis, and customize monitoring durations.

To configuring the Top N Queries settings:

1. From the **Query insights** page, Navigate to the **Configuration** tab from the **Query insights** page.

> 2. In the **Dashboards** window, select **Create**, then choose **Dashboard**.

Reviewer 2: What is this point to create dashboards???

Reviewer: I don't see this in my env, so I think we can delete it.

Author: I removed it.

> 4. Specify the monitoring **Window size**, determining the time duration for the query analysis.

Reviewer 2: > determining the time duration for the query analysis

This is very ambigious, could you update based on https://opensearch.org/docs/latest/observing-your-data/query-insights/top-n-queries/#configuring-the-window-size?

Reviewer 2: Also another issue regarding window size: https://github.com/opensearch-project/documentation-website/pull/9157#discussion_r1949845696

> 4. Specify the monitoring **Window size**, determining the time duration for the query analysis.

Reviewer: @ansjcy: Does this work? 

```suggestion
4. Specify the monitoring **Window size**, which determines the duration of time queries are collected for analysis. 
```

Reviewer 2: LGTM!

> 4. Toggle the **Enabled** setting to turn monitoring on or off for the selected metric.

Reviewer 2: -> Toggle the **Enabled** setting to turn Top N Queries feature on or off for the selected metric.

5. Enter the value of **N**, which defines the number of top queries to track.

> 6. Specify the monitoring **Window size**, determining the time duration for the query analysis.

Reviewer 2: Let's swap 5 and 6
for N, use:
Enter the value of **N**, which defines the number of top queries to track in each window.

### Top N queries grouping 

The **Top n queries group configuration settings** sets the grouping settings for queries.

You can set the following specific grouping attributes using these steps.

1. Select a grouping option under **Group By**, such as **Similarity**.
2. Select **Save**.
3. Check the **Statues for group by** panel to see if **Group by** criteria is enabled.

> 

Reviewer: Yes. Let's keep this for now and I'll make a PR for updating the UI text.

### Data export and retention

To configuring data export and retention, use the **Query insights export and data retention settings** panel. There, you can set the following settings:

1. Under **Exporter**, choose a destination where to export the insight data, such as **Local index**.
2. Set the **Delete After (days)** field with a data retention period.
3. Select **Save**.
4. In the **Statuses for data retention** panel, make sure that the **Exporter** setting is enabled.

### Configuration best practices

When configuring the query insights feature, remember the following best practices:

- Begin with a smaller value for N (count) and increase it based on your system's load.

> - For more comprehensive query insights, set a larger time window to analyze queries over a longer duration.

Reviewer: @ansjcy: In reference to your comment about window sizes. Does this language work for you?

> - For more comprehensive query insights, set a larger time window to analyze queries over a longer duration.

Reviewer 2: Yes, this works perfectly, thanks!

> - Begin with a smaller value for N (count) and increase it based on your system's load.

Reviewer 2: Let's also say use larger window if possible

Author: Added
